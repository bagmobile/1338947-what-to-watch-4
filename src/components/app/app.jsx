import React from 'react';
import Main from "../main/main.jsx";
import {PromoMovie} from "../../mocks/promo-movie.js";

const App = ({movies}) => {

  return (
    <Main
      movies={movies}
      promoMovie={PromoMovie}
    />
  );
};

App.propTypes = {
  movies: Main.propTypes.movies
};

export default App;
