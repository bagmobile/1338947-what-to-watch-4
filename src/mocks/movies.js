const mockReviews = [
  {
    author: `Kate Muir`,
    dateTime: `2020-03-07T14:34:25.500Z`,
    quote: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious
              Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in
              years.`,
    rating: 8.8
  },
  {
    author: `Bill Goodykoontz`,
    dateTime: `2020-03-07T14:34:25.500Z`,
    quote: `Anderson's films are too precious for some, but for those of us willing to lose
              ourselves in them, they're a delight. "The Grand Budapest Hotel" is no different, except that he has added
              a hint of gravitas to the mix, improving the recipe.`,
    rating: 8.9
  },
  {
    author: `Matthew Lickona`,
    dateTime: `2020-03-07T14:34:25.500Z`,
    quote: `The mannered, madcap proceedings are often delightful, occasionally silly, and
              here and there, gruesome and/or heartbreaking.`,
    rating: 3.9
  },
  {
    author: `Amanda Greever`,
    dateTime: `2020-03-07T14:34:25.500Z`,
    quote: `I didn't find it amusing, and while I can appreciate the creativity, it's an
              hour and 40 minutes I wish I could take back.`,
    rating: 5
  },
  {
    author: `Paula Fleri-Soler`,
    dateTime: `2020-03-07T14:34:25.500Z`,
    quote: `It is certainly a magical and childlike way of storytelling, even if the content
              is a little more adult.`,
    rating: 6.7
  },
  {
    author: `Paula Fleri-Soler`,
    dateTime: `2020-03-07T14:34:25.500Z`,
    quote: `It is certainly a magical and childlike way of storytelling, even if the content
              is a little more adult.`,
    rating: 8.9
  }
];

const movies = [
  {
    title: `Fantastic Beasts: The Crimes of Grindelwald`,
    img: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Crime`,
    year: 2010,
    ratingScore: 4,
    ratingCount: 100,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `1h 25m`,
    reviews: mockReviews
  },
  {
    title: `Bohemian Rhapsody`,
    img: `img/bohemian-rhapsody.jpg`,
    poster: `img/bohemian-rhapsody.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Documentary`,
    year: 2011,
    ratingScore: 4.4,
    ratingCount: 200,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `1h 35m`,
    reviews: mockReviews
  },
  {
    title: `Macbeth`,
    img: `img/macbeth.jpg`,
    poster: `img/macbeth.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Crime`,
    year: 1987,
    ratingScore: 6.7,
    ratingCount: 300,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `1h 50m`,
    reviews: mockReviews
  },
  {
    title: `Aviator`,
    img: `img/aviator.jpg`,
    poster: `img/aviator.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Comedies`,
    year: 2014,
    ratingScore: 4.5,
    ratingCount: 5,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `1h 35m`,
    reviews: mockReviews
  },
  {
    title: `We need to talk about Kevin`,
    img: `img/we-need-to-talk-about-kevin.jpg`,
    poster: `img/we-need-to-talk-about-kevin.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Comedies`,
    year: 1990,
    ratingScore: 2,
    ratingCount: 34,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `1h 15m`,
    reviews: mockReviews
  },
  {
    title: `Johnny English`,
    img: `img/johnny-english.jpg`,
    poster: `img/johnny-english.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Sci-Fi`,
    year: 1981,
    ratingScore: 0,
    ratingCount: 54,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `45m`,
    reviews: mockReviews
  },
  {
    title: `What We Do in the Shadows`,
    img: `img/what-we-do-in-the-shadows.jpg`,
    poster: `img/what-we-do-in-the-shadows.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Romance`,
    year: 2011,
    ratingScore: 9,
    ratingCount: 2,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `1h 15m`,
    reviews: mockReviews
  },
  {
    title: `Revenant`,
    img: `img/revenant.jpg`,
    poster: `img/revenant.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Comedies`,
    year: 2017,
    ratingScore: 7.7,
    ratingCount: 3,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `2h 35m`,
    reviews: mockReviews
  },
  {
    title: `Movie`,
    img: `img/revenant.jpg`,
    poster: `img/revenant.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Dramas`,
    year: 2013,
    ratingScore: 7.7,
    ratingCount: 3,
    text: [
      `Text1.`,
      `Text2`
    ],
    director: `Director`,
    starring: [`Star 1`, `Star 2`, `Star 3`, `Star 4`, `Star 5`, `Star 6`, `Star 7`, `Star 8`],
    duration: `2h 35m`,
    reviews: mockReviews
  }
];

export const MAX_GENRES_LIST_SIZE = 9;

export const DEFAULT_GENRE = `All genres`;

export const genres = [...movies.reduce((acc, movie) => {
  acc.add(movie.genre);
  return acc;
}, new Set([DEFAULT_GENRE]))].slice(0, MAX_GENRES_LIST_SIZE);


export default movies;
