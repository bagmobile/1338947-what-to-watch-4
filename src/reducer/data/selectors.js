import NameSpace from "../name-space";
import {DEFAULT_GENRE, MAX_GENRES_LIST_SIZE} from "../../consts";
import {createSelector} from "reselect";

const NAME_SPACE = NameSpace.DATA;

export const getMoviesList = (state) => state[NAME_SPACE].movies;

export const getReviews = (state) => state[NAME_SPACE].reviews;

export const getPromoMovie = (state) => state[NAME_SPACE].promoMovie;

export const getFavoriteMovies = (state) => state[NAME_SPACE].favoriteMovies;

export const getMovieById = (state, id) => getMoviesList(state).find((movie) => movie.id === id);

export const getMoviesGenres = createSelector(
    getMoviesList,
    (movies) => [...movies.reduce((acc, movie) => {
      acc.add(movie.genre);
      return acc;
    }, new Set([DEFAULT_GENRE]))].slice(0, MAX_GENRES_LIST_SIZE)
);

export const isFetchingMovies = (state) => getMoviesList(state).length === 0;

export const isFetchingPromoMovie = (state) => getPromoMovie(state) === null;
