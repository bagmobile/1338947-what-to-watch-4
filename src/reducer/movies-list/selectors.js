import {createSelector} from "reselect";
import NameSpace from "../name-space";
import {DEFAULT_GENRE, DEFAULT_MOVIE_LIST_SIZE, MOVIE_LIST_BY_GENRE_SIZE} from "../../consts";

const getCurrentMovieListSize = (state) => state[NameSpace.MOVIES_LIST].currentMovieListSize;

const getMoviesList = (state) => {
  return state[NameSpace.DATA].movies;
};

export const getActiveGenre = (state) => state[NameSpace.MOVIES_LIST].activeGenre;

export const getMoviesLike = (state, activeMovie, size = MOVIE_LIST_BY_GENRE_SIZE) => state[NameSpace.DATA].movies
  .filter((movie) => movie.genre === activeMovie.genre)
  .filter((movie) => movie.id !== activeMovie.id)
  .slice(0, size);

export const getFilteredMoviesByGenre = createSelector(
    [getMoviesList, getActiveGenre],
    (moviesList, activeGenre) => moviesList.filter((movie) => activeGenre === DEFAULT_GENRE
    || movie.genre === activeGenre
    ));

export const getPartFilteredMoviesByGenre = createSelector(
    [getFilteredMoviesByGenre, getCurrentMovieListSize],
    (moviesByGenre, listSize) => moviesByGenre.slice(0, listSize)
);

export const getVisibilityShowMoreButton = createSelector(
    [getFilteredMoviesByGenre, getCurrentMovieListSize],
    (moviesByGenre, listSize) => (moviesByGenre.length >= DEFAULT_MOVIE_LIST_SIZE) && (listSize < moviesByGenre.length)
);
