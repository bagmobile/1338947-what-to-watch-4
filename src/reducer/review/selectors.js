import NameSpace from "../name-space";
import {FetchingStatus} from "../../consts";

const NAME_SPACE = NameSpace.REVIEW;

export const getReviews = (state) => state[NAME_SPACE].reviews;

export const getMessage = (state) => state[NAME_SPACE].message;

export const isReviewFetching = (state) => state[NAME_SPACE].reviewFetchingStatus === FetchingStatus.IN_PROGRESS;

export const isSuccessReviewsFetching = (state) => state[NAME_SPACE].reviewFetchingStatus === FetchingStatus.SUCCESS;

export const isErrorReviewFetching = (state) => state[NAME_SPACE].reviewFetchingStatus === FetchingStatus.ERROR;

export const isAddReviewFetching = (state) => state[NAME_SPACE].addReviewFetchingStatus === FetchingStatus.IN_PROGRESS;

export const isSuccessAddReviewsFetching = (state) => state[NAME_SPACE].addReviewFetchingStatus === FetchingStatus.SUCCESS;

export const isErrorAddReviewFetching = (state) => state[NAME_SPACE].addReviewFetchingStatus === FetchingStatus.ERROR;
