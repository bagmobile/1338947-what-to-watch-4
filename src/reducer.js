import movies, {DEFAULT_GENRE} from "./mocks/movies";
import {extend} from "./utils/util";
import {PromoMovie} from "./mocks/promo-movie";

const initialState = {
  movies,
  activeMovie: null,
  promoMovie: PromoMovie,
  activeGenre: DEFAULT_GENRE
};

const ActionType = {
  CHANGE_GENRE: `CHANGE GENRE`,
  FILTER_MOVIES: `FILTER MOVIES`,
  SELECT_MOVIE: `SELECT MOVIE`
};

const filterMoviesByGenre = (genre, removedMovie) => {
  if (genre === DEFAULT_GENRE) {
    return movies;
  }
  return movies.filter((movie) => {
    if (removedMovie && (removedMovie === movie)) {
      return false;
    }
    return movie.genre === genre;
  });
};

const ActionCreator = {
  changeGenre: (activeGenre = DEFAULT_GENRE) => ({
    type: ActionType.CHANGE_GENRE,
    payload: activeGenre
  }),
  filterMovies: (activeGenre, removeMovie = null) => ({
    type: ActionType.FILTER_MOVIES,
    payload: filterMoviesByGenre(activeGenre, removeMovie)
  }),
  selectMovie: (activeMovie = null) => ({
    type: ActionType.SELECT_MOVIE,
    payload: activeMovie
  })
};

const reducer = (state = initialState, action) => {

  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      return extend(state, {activeGenre: action.payload});
    case ActionType.FILTER_MOVIES:
      return extend(state, {movies: action.payload});
    case ActionType.SELECT_MOVIE:
      return extend(state, {activeMovie: action.payload});
  }
  return state;
};

export {reducer, ActionType, ActionCreator, initialState};
