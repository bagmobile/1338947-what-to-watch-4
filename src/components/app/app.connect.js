import {Operation as DataOperation} from "../../reducer/data/data";
import {Operation as UserOperation} from "../../reducer/user/user";
import {connect} from "react-redux";
import App from "../../components/app/app";

const mapDispatchToProps = (dispatch) => ({
  init: () => {
    dispatch(UserOperation.checkAuth());
    dispatch(DataOperation.loadMovies());
    dispatch(DataOperation.loadPromoMovie());
  },
});

export default connect(null, mapDispatchToProps)(App);
