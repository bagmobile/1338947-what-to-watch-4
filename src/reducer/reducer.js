import {combineReducers} from "redux";
import NameSpace from "./name-space";
import {reducer as data} from "./data/data";
import {reducer as moviesList} from "./movies-list/movies-list";
import {reducer as user} from "./user/user";
import {reducer as review} from "./review/review";


export default combineReducers({
  [NameSpace.DATA]: data,
  [NameSpace.MOVIES_LIST]: moviesList,
  [NameSpace.USER]: user,
  [NameSpace.REVIEW]: review
});
