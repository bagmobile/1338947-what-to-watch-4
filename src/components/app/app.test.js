import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {App} from "./app.jsx";
import {Provider} from "react-redux";
import {initialState} from "../../reducer";
import {genres} from "../../mocks/movies";
import {getPromoMovie} from "../../selectors";

const mockStore = configureStore([]);

describe(`App component`, () => {

  it(`Render`, () => {
    const store = mockStore(initialState);
    const tree = renderer.create(
        <Provider store={store}>
          <App
            genres={genres}
            promoMovie={getPromoMovie(initialState)}
          />
        </Provider>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
