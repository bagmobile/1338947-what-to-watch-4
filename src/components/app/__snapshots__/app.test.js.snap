// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Render 1`] = `
Array [
  <section
    className="movie-card"
    style={
      Object {
        "backgroundColor": "#A6B7AC",
      }
    }
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Bohemian Rhapsody"
        src="https://htmlacademy-react-3.appspot.com/wtw/static/film/background/gangs_of_new_york.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster "
        >
          <img
            alt="Bohemian Rhapsody poster"
            height="327"
            src="img/bohemian-rhapsody.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Bohemian Rhapsody
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Crime
            </span>
            <span
              className="movie-card__year"
            >
              2011
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/player/1"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item false"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Crime
          </a>
        </li>
        <li
          className="catalog__genres-item false"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Comedies
          </a>
        </li>
        <li
          className="catalog__genres-item false"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Dramas
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            className="small-movie-card__link"
            href="/movies/1"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="175"
                loop={true}
                poster="img/bohemian-rhapsody.jpg"
                src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Bohemian Rhapsody
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            className="small-movie-card__link"
            href="/movies/2"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="175"
                loop={true}
                poster="img/bohemian-rhapsody.jpg"
                src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Bohemian Rhapsody
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            className="small-movie-card__link"
            href="/movies/3"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="175"
                loop={true}
                poster="img/bohemian-rhapsody.jpg"
                src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Bohemian Rhapsody
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            className="small-movie-card__link"
            href="/movies/4"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="175"
                loop={true}
                poster="img/bohemian-rhapsody.jpg"
                src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Bohemian Rhapsody
              </span>
            </h3>
          </a>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
